pkg/apis/networking/validation/validation_test.go

```diff
@@ -1853,3 +1853,209 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv6CIDR(nil)),
+		"valid DualStack ClusterCIDRConfig with no NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(nil)),
+		"32 - IPv4.PerNodeMaskSize == 128 - IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 24)),
+			setIPv6CIDR(makeCIDRConfig("fd00::/112", 120))),
+		"valid NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(validNodeSelector)),
+	}
+
+	// Success cases are expected to pass validation.
+
+	for k, v := range successCases {
+		if errs := ValidateClusterCIDRConfig(v); len(errs) != 0 {
+			t.Errorf("Expected success for test '%s', got %v", k, errs)
+		}
+	}
+
+	invalidNodeSelector := makeNodeSelector("NoUppercaseOrSpecialCharsLike=Equals", api.NodeSelectorOpIn, []string{"bar"})
+
+	errorCases := map[string]*networking.ClusterCIDRConfig{
+		// Config test.
+		"nil spec.IPv4 and spec.IPv6": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(nil), setIPv6CIDR(nil)),
+		"invalid spec.NodeSelector": makeClusterCIDRConfigCustom(
+			setNodeSelector(invalidNodeSelector)),
+
+		// IPv4 tests.
+		"empty spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("", 24))),
+		"invalid spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("test", 24))),
+		"valid IPv6 CIDR in spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("fd00::/120", 24))),
+		"invalid spec.IPv4.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 100))),
+		"invalid spec.IPv4.PerNodeMaskSize < CIDR Mask": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 12))),
+
+		// IPv6 tests.
+		"empty spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("", 120))),
+		"invalid spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("testv6", 120))),
+		"valid IPv4 CIDR in spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("10.2.0.0/16", 24))),
+		"invalid spec.IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 1000))),
+		"invalid spec.IPv6.PerNodeMaskSize < CIDR Mask": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 119))),
+
+		// DualStack test.
+		"32 - IPv4.PerNodeMaskSize != 128 - IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 24)),
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 112))),
+	}
+
+	// Error cases are not expected to pass validation.
+	for testName, ccc := range errorCases {
+		if errs := ValidateClusterCIDRConfig(ccc); len(errs) == 0 {
+			t.Errorf("Expected failure for test: %s", testName)
+		}
+	}
+}
+
+func TestValidateClusterConfigUpdate(t *testing.T) {
+	oldCCC := makeValidClusterCIDRConfig()
+
+	// Tweaks used below.
+	setIPv4CIDR := func(cidrConfig *networking.CIDRConfig) cccTweak {
```

Comment:
why not move all the tweaks outside the function. They are shared between create/update validation tests.

Also i have not read them yet but you will also need the make/tweak functions for strategy testing so better move them to a package

Reply:
Removed the tweaks, was confusing, now creating a new ClusterCIDR object for every  test